File: xcloak-agent-mobile/test/deception_test.dart
// Tests for canary token and honeyport API shapes used by the mobile agent.
// The mobile agent can trip canary tokens embedded in documents and URLs
// that are distributed to mobile devices.
//
// Run with: flutter test

import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Deception (canary tokens & honeyports)', () {
    test('canary token record includes required fields', () {
      final token = {
        'id': 1,
        'token_type': 'url',
        'name': 'S3 Bucket Canary',
        'token_value': 'c-abc123',
        'is_active': true,
        'trip_count': 0,
      };

      for (final key in ['id', 'token_type', 'name', 'token_value', 'is_active']) {
        expect(token.containsKey(key), isTrue,
            reason: 'canary token must include "$key" for the deception UI');
      }
    });

    test('toggle body uses is_active bool field', () {
      final body = {'is_active': false};
      expect(body['is_active'], isA<bool>(),
          reason: 'toggle PATCH body must use is_active as bool');
    });

    test('canary token and honeyport URL patterns are correct', () {
      String tokenUrl(int id) => '/api/canary/tokens/$id';
      String tokenToggle(int id) => '/api/canary/tokens/$id/toggle';
      String honeyportUrl(int id) => '/api/honeyports/$id';

      expect(tokenUrl(3), equals('/api/canary/tokens/3'));
      expect(tokenToggle(3), equals('/api/canary/tokens/3/toggle'));
      expect(honeyportUrl(7), equals('/api/honeyports/7'));
    });
  });
}

File: xcloak-agent-desktop/agent/deception_test.go
package agent

import (
	"encoding/json"
	"fmt"
	"testing"
)

// TestCanaryToken_URLPatterns verifies correct URL construction for canary
// token CRUD and toggle operations.
func TestCanaryToken_URLPatterns(t *testing.T) {
	cases := []struct {
		got  string
		want string
	}{
		{fmt.Sprintf("/api/canary/tokens/%d", 3), "/api/canary/tokens/3"},
		{fmt.Sprintf("/api/canary/tokens/%d/toggle", 3), "/api/canary/tokens/3/toggle"},
	}
	for _, c := range cases {
		if c.got != c.want {
			t.Errorf("URL = %q, want %q", c.got, c.want)
		}
	}
}

// TestCanaryTokenPayload_Fields verifies token record includes required fields.
func TestCanaryTokenPayload_Fields(t *testing.T) {
	raw := `{"id":1,"token_type":"url","name":"S3 Bucket Canary","token_value":"c-abc123","is_active":true,"trip_count":0}`
	var token map[string]any
	if err := json.Unmarshal([]byte(raw), &token); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, key := range []string{"id", "token_type", "name", "token_value", "is_active"} {
		if _, ok := token[key]; !ok {
			t.Errorf("canary token payload missing field %q", key)
		}
	}
}

// TestHoneyportPayload_Fields verifies honeyport record includes required
// fields needed to display and manage ports.
func TestHoneyportPayload_Fields(t *testing.T) {
	raw := `{"id":1,"port":4444,"protocol":"tcp","description":"test","is_active":true}`
	var hp map[string]any
	if err := json.Unmarshal([]byte(raw), &hp); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, key := range []string{"id", "port", "protocol", "is_active"} {
		if _, ok := hp[key]; !ok {
			t.Errorf("honeyport payload missing field %q", key)
		}
	}
}
